modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Divide.kt

```diff
@@ -0,0 +1,318 @@
+   *   }.fix()
+   *
+   *   val user = User("John", 31)
+   *
+   *   val result = userSerializer.func(user)
+   *   //sampleEnd
+   *   println(result)
+   * }
+   * ```
+   */
+  fun <A, B, Z> divide(fa: Kind<F, A>, fb: Kind<F, B>, f: (Z) -> Tuple2<A, B>): Kind<F, Z>
+
+  fun <A, B> Kind<F, A>.product(other: Kind<F, B>): Kind<F, Tuple2<A, B>> =
+    divide(this, other, ::identity)
+
+  fun <A, B, C> Kind<F, Tuple2<A, B>>.product(
+    other: Kind<F, C>,
+    dummy: Unit = Unit
+  ): Kind<F, Tuple3<A, B, C>> =
+    divide(this, other) { Tuple2(it.a toT it.b, it.c) }
+
+  fun <A, B, C, D> Kind<F, Tuple3<A, B, C>>.product(
+    other: Kind<F, D>,
+    dummy: Unit = Unit,
+    dummy2: Unit = Unit
+  ): Kind<F, Tuple4<A, B, C, D>> =
+    divide(this, other) { Tuple2(Tuple3(it.a, it.b, it.c), it.d) }
+
+  fun <A, B, C, D, E> Kind<F, Tuple4<A, B, C, D>>.product(
+    other: Kind<F, E>,
+    dummy: Unit = Unit,
+    dummy2: Unit = Unit,
+    dummy3: Unit = Unit
+  ): Kind<F, Tuple5<A, B, C, D, E>> =
+    divide(this, other) { Tuple2(Tuple4(it.a, it.b, it.c, it.d), it.e) }
+
+  fun <A, B, C, D, E, FF> Kind<F, Tuple5<A, B, C, D, E>>.product(
+    other: Kind<F, FF>,
+    dummy: Unit = Unit,
+    dummy2: Unit = Unit,
+    dummy3: Unit = Unit,
+    dummy4: Unit = Unit
+  ): Kind<F, Tuple6<A, B, C, D, E, FF>> =
+    divide(this, other) { Tuple2(Tuple5(it.a, it.b, it.c, it.d, it.e), it.f) }
+
+  fun <A, B, C, D, E, FF, G> Kind<F, Tuple6<A, B, C, D, E, FF>>.product(
+    other: Kind<F, G>,
+    dummy: Unit = Unit,
+    dummy2: Unit = Unit,
+    dummy3: Unit = Unit,
+    dummy4: Unit = Unit,
+    dummy5: Unit = Unit
+  ): Kind<F, Tuple7<A, B, C, D, E, FF, G>> =
+    divide(this, other) { Tuple2(Tuple6(it.a, it.b, it.c, it.d, it.e, it.f), it.g) }
+
+  fun <A, B, C, D, E, FF, G, H> Kind<F, Tuple7<A, B, C, D, E, FF, G>>.product(
+    other: Kind<F, H>,
+    dummy: Unit = Unit,
+    dummy2: Unit = Unit,
+    dummy3: Unit = Unit,
+    dummy4: Unit = Unit,
+    dummy5: Unit = Unit,
+    dummy6: Unit = Unit
+  ): Kind<F, Tuple8<A, B, C, D, E, FF, G, H>> =
+    divide(this, other) { Tuple2(Tuple7(it.a, it.b, it.c, it.d, it.e, it.f, it.g), it.h) }
+
+  fun <A, B, C, D, E, FF, G, H, I> Kind<F, Tuple8<A, B, C, D, E, FF, G, H>>.product(
+    other: Kind<F, I>,
+    dummy: Unit = Unit,
+    dummy2: Unit = Unit,
+    dummy3: Unit = Unit,
+    dummy4: Unit = Unit,
+    dummy5: Unit = Unit,
+    dummy6: Unit = Unit,
+    dummy7: Unit = Unit
+  ): Kind<F, Tuple9<A, B, C, D, E, FF, G, H, I>> =
+    divide(this, other) { Tuple2(Tuple8(it.a, it.b, it.c, it.d, it.e, it.f, it.g, it.h), it.i) }
+
+  fun <A, B, C, D, E, FF, G, H, I, J> Kind<F, Tuple9<A, B, C, D, E, FF, G, H, I>>.product(
+    other: Kind<F, J>,
+    dummy: Unit = Unit,
+    dummy2: Unit = Unit,
+    dummy3: Unit = Unit,
+    dummy4: Unit = Unit,
+    dummy5: Unit = Unit,
+    dummy6: Unit = Unit,
+    dummy7: Unit = Unit,
+    dummy8: Unit = Unit
+  ): Kind<F, Tuple10<A, B, C, D, E, FF, G, H, I, J>> =
```

Comment:
RIP IntelliJ

Reply:
I've done worse with a small project regarding generators for prop testing (generated ~1k lines of Tuple code for generating tuples and shrinkers for them) :) IntelliJ takes a while to load that file :D